src/main/java/eu/europeana/metis/sandbox/service/metrics/MetricsServiceImpl.java

```diff
@@ -0,0 +1,156 @@
+
+/**
+ * Metrics Service implementation class
+ */
+@Service
+public class MetricsServiceImpl implements MetricsService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetricsServiceImpl.class);
+  private static final String METRICS_NAMESPACE = "sandbox.metrics.dataset";
+  public static final String BASE_UNIT_RECORD = "record";
+  public static final String BASE_UNIT_DATASET = "Dataset";
+  private final RecordRepository recordRepository;
+  private final RecordLogRepository recordLogRepository;
+  private final DatasetProblemPatternRepository problemPatternRepository;
+  private final MeterRegistry meterRegistry;
+  private List<DatasetStatistic> datasetStatistics;
+  private List<StepStatistic> stepStatistics;
+  private List<DatasetProblemPatternStatistic> problemPatternStatistics;
+
+  public MetricsServiceImpl(
+      RecordRepository recordRepository,
+      RecordLogRepository recordLogRepository,
+      DatasetProblemPatternRepository problemPatternRepository,
+      MeterRegistry meterRegistry) {
+    this.recordRepository = recordRepository;
+    this.recordLogRepository = recordLogRepository;
+    this.problemPatternRepository = problemPatternRepository;
+    this.meterRegistry = meterRegistry;
+    initMetrics();
+  }
+
+  @Scheduled(cron = "${" + METRICS_NAMESPACE + ".frequency:*/5 * * * * *")
+  @Override
+  public void processMetrics() {
+    datasetStatistics = recordRepository.getMetricDatasetStatistics();
+    stepStatistics = recordLogRepository.getMetricStepStatistics();
+    problemPatternStatistics = problemPatternRepository.getMetricProblemPatternStatistics();
+    LOGGER.debug("metrics report retrieval");
+  }
+
+  Long getDatasetCount() {
+    if (datasetStatistics == null || datasetStatistics.isEmpty()) {
+      return 0L;
+    } else {
+      return (long) datasetStatistics.size();
+    }
+  }
+
+  Long getTotalRecords() {
+    if (datasetStatistics == null || datasetStatistics.isEmpty()) {
+      return 0L;
+    } else {
+      return datasetStatistics.stream().mapToLong(DatasetStatistic::getCount).sum();
+    }
+  }
+
+  Long getTotalOcurrences(ProblemPatternId problemPatternId) {
+    if (problemPatternStatistics == null || problemPatternStatistics.isEmpty()) {
+      return 0L;
+    } else {
+      return problemPatternStatistics
+          .stream()
+          .filter(pattern -> pattern.getPatternId().equals(problemPatternId.name()))
+          .mapToLong(DatasetProblemPatternStatistic::getCount).sum();
+    }
+  }
+
+  Long getTotalRecords(Step step, Status status) {
+    if (stepStatistics == null || stepStatistics.isEmpty()) {
+      return 0L;
+    } else {
+      return stepStatistics.stream()
+                           .filter(stepStatistic -> stepStatistic.getStep().equals(step) &&
+                               stepStatistic.getStatus().equals(status))
+                           .mapToLong(StepStatistic::getCount).sum();
+    }
+  }
+
+  void initMetrics() {
```

Comment:
Either make this final or perhaps don't use the interface implements because calling it from the constructor with inheritance can be dangerous.

Reply:
Yes. thanks. I made it private. so only via the constructor is called this method.